source/agora/consensus/validation/Block.d

```diff
@@ -112,6 +118,268 @@ public string isInvalidReason (const ref Block block, in ulong prev_height,
+        is invalid otherwise.
+
+*******************************************************************************/
+
+private string isValidGenesisBlock (const ref Block genesisBlock) nothrow @safe
+{
+    if (genesisBlock.header.height != 0)
+        return "GenesisBlock: The height of the block is not 0";
+
+    if (genesisBlock.header.prev_block != Hash.init)
+        return "GenesisBlock: Header.prev_block is not empty";
+
+    if (genesisBlock.txs.length == 0)
+        return "GenesisBlock: Transaction(s) is empty";
+
+    if (!genesisBlock.txs.isSorted())
+        return "GenesisBlock: Transactions are not sorted";
+
+    if (genesisBlock.txs.length > Block.TxsInBlock)
+        return "GenesisBlock: The number of transactions is out of bounds";
+
+    UTXOSetValue[Hash] utxo_set;
+    foreach (const ref tx; genesisBlock.txs)
+    {
+        if (!(tx.type == TxType.Payment || tx.type == TxType.Freeze))
+            return "GenesisBlock: Transaction type allows only "
+                ~ "Payment and Freeze";
+
+        if (tx.inputs.length != 1)
+            return "GenesisBlock: Transaction must have only one input";
+
+        if (tx.inputs[0] != Input.init)
+             return "GenesisBlock: Not empty a input in the transaction";
+
+        if (tx.outputs.length == 0)
+            return "GenesisBlock: No output(s) in the transaction";
+
+        Hash tx_hash = tx.hashFull();
+        foreach (idx, const ref output; tx.outputs)
+        {
+            // disallow negative amounts
+            if (!output.value.isValid())
+                return "GenesisBlock: Output(s) overflow or underflow"
+                    ~ "in the transaction";
+
+            // disallow 0 amount
+            if (output.value == Amount(0))
+                return "GenesisBlock: Value of output is 0"
+                    ~ "in the transaction";
+
+            const UTXOSetValue utxo_value = {
+                unlock_height: 0,
+                type: tx.type,
+                output: output
+            };
+            utxo_set[UTXOSet.getHash(tx_hash, idx)] = utxo_value;
+        }
+    }
+
+    Hash[] merkle_tree;
+    if (genesisBlock.header.merkle_root !=
+        Block.buildMerkleTree(genesisBlock.txs, merkle_tree))
+        return "GenesisBlock: Merkle root does not match header's";
+
+    Set!Hash used_utxos;
+
+    bool findUTXO (Hash hash, size_t index, out UTXOSetValue value)
+        nothrow @safe
+    {
+        Hash utxo_hash;
+
+        if (index == size_t.max)
+            utxo_hash = hash;
+        else
+            utxo_hash = UTXOSet.getHash(hash, index);
+
+        if (utxo_hash in used_utxos)
+            return false;  // double-spend
+
+        if (auto ptr = utxo_hash in utxo_set)
+        {
+            value = *ptr;
+            used_utxos.put(utxo_hash);
+            return true;
+        }
+        return false;
+    }
+
+    // If there are no enrollments, return them here early.
```

Comment:
Move this check above where `used_utxos` is defined

Reply:
I moved it.